File: 1/1.hs
 -- 1:1 conversion of 1.anubis to Haskell --

 -------------------------------------------
 --    by Julien Verneuil, 12/06/2015     --
 -------------------------------------------

 --       ghc 1.hs -o projecteuler_1      --

import System.Environment
import Debug.Trace
import Data.List
import Data.Maybe
import Text.Read

integer_sum :: Integer -> Integer
integer_sum n = (n * (n + 1)) `quot` 2

signed_lcm_list :: [Integer] -> [Integer]
signed_lcm_list l = 
    if null l then []
    else
        let h = head l
            t = tail l in
        if null t then [h]
        else
            let p = signed_lcm_list(t) in
            [h] ++ map (\x -> -(lcm h x)) p ++ p

print_ignore_message :: String -> a -> a
print_ignore_message s = trace("Parameter \"" ++ s ++ "\" ignored, not a natural number, you should fix this.\n")

to_integer :: String -> Maybe Integer
to_integer s = 
    case readMaybe s of
        Nothing ->
            print_ignore_message s
            Nothing
        Just n -> 
            if n <= 0 then
                print_ignore_message s
                Nothing
            else
                Just n

usage :: String
usage = "Usage:\n    projecteuler_1 max n1 n2 n3 ...\n\nExample:\n    projecteuler_1 999 3 5"

main = do
    args <- getArgs

    let args_as_int = mapMaybe ( to_integer ) args
        
    if null args_as_int then putStrLn usage
    else do
        let max         = head args_as_int 
            numbers     = tail args_as_int

        if null numbers then putStrLn usage
        else do
            let slcml       = signed_lcm_list(numbers)
                result      = sum(map (\x -> integer_sum(max `quot` abs(x)) * x) slcml)

            putStrLn("The sum of all the multiples of " ++ intercalate " or " (tail args) ++ " below " ++ show(max + 1) ++ " is " ++ show result ++ ".")

File: 1/1.py
# 1:1 conversion of 1.anubis to Python (2.x)

# by Julien Verneuil, 14/06/2015

from fractions import gcd
from sys import argv
from sys import exit

def print_usage():
    print("Usage:\n    python 1.py max n1 n2 n3 ...\n\nExample:\n    python 1.py 999 3 5")

def print_ignore_message(s):
    print("Parameter \"" + s + "\" ignored, not a natural number, you should fix this.")

def lcm(a, b):
    return (a * b // abs(gcd(a, abs(b))))

def n_sum(n):
    return (n * (n + 1)) // 2

def signed_lcm_list(l):
    if not l:
        return []
    head, tail = l[0], l[1:]
    if not tail:
        return [head]
    p = signed_lcm_list(tail)
    new_l = [head]
    new_l.extend([-lcm(head, x) for x in p])
    new_l.extend(p)
    return new_l

args_as_int = []

for arg in argv[1:]:
    try:
        n = int(arg)
        if n <= 0:
            raise ValueError()

        args_as_int.append(int(arg))
    except ValueError:
        print_ignore_message(arg)

if len(args_as_int) <= 1:
    print_usage()
    exit(0)

max = args_as_int.pop(0)

slcml = signed_lcm_list(args_as_int)
result = []

for n in slcml:
    result.append(n_sum(max // n) * n)

result = sum(result)

print("The sum of all the multiples of " + " or ".join(argv[2:]) + " below " + str(max + 1) + " is " + str(result))
